Assets/MixedRealityToolkit.Services/DiagnosticsSystem/MixedRealityToolkitVisualProfiler.cs

```diff
@@ -38,109 +47,135 @@ public bool IsVisible
+        private Quaternion windowVerticalRotation;
+        private Quaternion windowVerticalRotationInverse;
+
+        private Matrix4x4[] frameInfoMatricies;
+        private Vector4[] frameInfoColors;
+        private MaterialPropertyBlock frameInfoPropertyBlock;
+        private int colorID;
+        private int parentMatrixID;
         private int frameCount;
-        private Stopwatch stopwatch = new Stopwatch();
-        private string[] frameRateStrings;
-        private string displayedDecimalFormat;
+        private System.Diagnostics.Stopwatch stopwatch = new System.Diagnostics.Stopwatch();
+        private FrameTiming[] frameTimings = new FrameTiming[maxFrameTimings];
+        private string[] cpuFrameRateStrings;
+        private string[] gpuFrameRateStrings;
+        private char[] stringBuffer = new char[maxStringLength];
 
         private ulong memoryUsage;
         private ulong peakMemoryUsage;
         private ulong limitMemoryUsage;
 
-        private MaterialPropertyBlock propertyBlockFrameTarget;
-        private MaterialPropertyBlock propertyBlockFrameMissed;
-
-        private StringBuilder stringBuilder = new StringBuilder(32);
-
-        [SerializeField]
-        [HideInInspector]
+        // Rendering resources.
+        [SerializeField, HideInInspector]
         private Material defaultMaterial;
+        [SerializeField, HideInInspector]
+        private Material defaultInstancedMaterial;
         private Material backgroundMaterial;
         private Material foregroundMaterial;
         private Material textMaterial;
+        private Mesh quadMesh;
 
         private void Reset()
         {
-            if (!defaultMaterial)
+            if (defaultMaterial == null)
             {
                 defaultMaterial = new Material(Shader.Find("Hidden/Internal-Colored"));
                 defaultMaterial.SetFloat("_ZWrite", 0.0f);
-                defaultMaterial.SetFloat("_ZTest", 0.0f);
+                defaultMaterial.SetFloat("_ZTest", (float)UnityEngine.Rendering.CompareFunction.Disabled);
                 defaultMaterial.renderQueue = 5000;
             }
 
+            if (defaultInstancedMaterial == null)
+            {
+                Shader defaultInstancedShader = Shader.Find("Hidden/Instanced-Colored");
+
+                if (defaultInstancedShader != null)
+                {
+                    defaultInstancedMaterial = new Material(defaultInstancedShader);
+                    defaultInstancedMaterial.enableInstancing = true;
+                    defaultInstancedMaterial.SetFloat("_ZWrite", 0.0f);
+                    defaultInstancedMaterial.SetFloat("_ZTest", (float)UnityEngine.Rendering.CompareFunction.Disabled);
+                    defaultInstancedMaterial.renderQueue = 5000;
+                }
+                else
+                {
+                    Debug.LogWarning("A shader supporting instancing could not be found for the VisualProfiler, falling back to traditional rendering.");
```

Comment:
Maybe add a warning "This may impact performance"? Otherwise it is not clear why falling back to traditional rendering is a warning.

Reply:
Great idea, added!